libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp

```diff
@@ -0,0 +1,486 @@
+}
+
+// update the _rp_state of the sensor
+void AP_Proximity_RPLidarA2::update(void)
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // initialise sensor if necessary
+    if(!_initialised)
+    {
+        _initialised = initialise();    //returns true if everything initialized properly
+    }
+
+
+    // if LIDAR in known state
+    if (_initialised) {
+        get_readings();
+    }
+
+    // check for timeout and set health status
+    if ((_last_distance_received_ms == 0) || (AP_HAL::millis() - _last_distance_received_ms > COMM_ACTIVITY_TIMEOUT_MS)) {
+        set_status(AP_Proximity::Proximity_NoData);
+#if RP_DEBUG_LEVEL > 0
+        GCS_MAVLINK::send_text(MAV_SEVERITY_INFO, "LIDAR NO DATA");
+#endif
+    } else {
+        set_status(AP_Proximity::Proximity_Good);
+    }
+}
+
+// get maximum distance (in meters) of sensor
+float AP_Proximity_RPLidarA2::distance_max() const
+{
+    return 16.0f;  //16m max range RPLIDAR2
+}
+
+// get minimum distance (in meters) of sensor
+float AP_Proximity_RPLidarA2::distance_min() const
+{
+    return 0.20f;  //20cm min range RPLIDAR2
+}
+
+
+bool AP_Proximity_RPLidarA2::initialise()
+{
+    // initialise sectors
+    if (!_sector_initialised) {
+        init_sectors();
+        return false;
+    }
+    if (!_initialised) {
+        reset_rplidar();            // set to a known state
+#if RP_DEBUG_LEVEL > 0
+        GCS_MAVLINK::send_text(MAV_SEVERITY_INFO, "LIDAR initialised");
+#endif
+        return true;
+    }
+
+    return true;
+}
+
+void AP_Proximity_RPLidarA2::reset_rplidar()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+#if FTDI_DEBUG
+    _uart->write(0x32); //SP as padding for ftdi cable
+    _uart->write(0x32);
+    _uart->write(0x32); //SP as padding for ftdi cable
+    _uart->write(0x32);
+    _uart->write(0x32); //SP as padding for ftdi cable
+    _uart->write(0x32);
+    _uart->write(0x32); //SP as padding for ftdi cable
+    _uart->write(0x32);
+#endif
+    uint8_t tx_buffer[2] = {RPLIDAR_PREAMBLE, RPLIDAR_CMD_RESET};
+    _uart->write(tx_buffer, 2);
+    _resetted = true;                   ///< be aware of extra 63 bytes coming after reset containing FW information
+#if RP_DEBUG_LEVEL > 0
+    GCS_MAVLINK::send_text(MAV_SEVERITY_INFO, "LIDAR reset");
+#endif
+    hal.scheduler->delay(RESET_RPA2_WAIT_MS);                      ///< wait at least 8ms after sending new request
+    //hal.scheduler->delay_microseconds(PROXIMITY_RPA2_TIMEOUT_MS * 1000);
+    _last_reset_ms =  AP_HAL::millis();
+    _rp_state = rp_resetted;
+
+}
+
+// initialise sector angles using user defined ignore areas, left same as SF40C
+// TODO: adapt to RPLIDAR A2 measurements
+void AP_Proximity_RPLidarA2::init_sectors()
+{
+    // use defaults if no ignore areas defined
+    uint8_t ignore_area_count = get_ignore_area_count();
+    if (ignore_area_count == 0) {
+        _sector_initialised = true;
+        return;
```

Comment:
Again that is the same code as AP_Proximity_LightWareSF40C.cpp
That again needs to be fixed in there as well.

Reply:
we need the special check so that if the "ignore_area_count" is zero, we leave the _num_sectors and _boundary array at their defaults which is 8 sectors of 45degrees each.  Without the special check the _num_sectors would be set to zero.